File: bridge-android/core/src/main/kotlin/mx/acstechnology/doomyvision/core/Wearables.kt
package mx.acstechnology.doomyvision.core

/**
 * Abstracción de los tres modos de operación (sección 6):
 *   MODE A — REAL_GLASSES (Ray-Ban reales vía Meta Wearables DAT)
 *   MODE B — MOCK_DEVICE (Meta MockDeviceKit)
 *   MODE C — PHONE (cámara/mic/bocinas del teléfono)
 *
 * `WearablesManager` es la interfaz que la UI y el resto del Bridge
 * consumen. La implementación real (MODE A) vive en app/ y envuelve
 * MWDATCore/MWDATCamera (ver docs/DOOMY_VISION_ARCHITECTURE.md) — no
 * puede compilarse en este entorno (requiere Android SDK + acceso al
 * Developer Preview de Meta). MockWearablesManager (MODE B, esta clase)
 * SÍ es real y probada: reproduce el flujo documentado de MockDeviceKit
 * (pair -> power on -> don -> unfold -> session ready -> camera available).
 */
enum class DeviceMode { REAL_GLASSES, MOCK_DEVICE, PHONE }

enum class WearableConnectionState { DISCONNECTED, PAIRING, PAIRED, POWERED_ON, WORN, SESSION_READY }

data class WearableStatus(
    val connectionState: WearableConnectionState,
    val cameraAvailable: Boolean,
    val batteryPercent: Int? = null,
)

interface WearablesManager {
    val mode: DeviceMode
    suspend fun pair(): WearableStatus
    suspend fun powerOn(): WearableStatus
    suspend fun don(): WearableStatus /** el usuario se pone los lentes */
    suspend fun unfold(): WearableStatus /** desdoblar — requerido antes de iniciar sesión */
    suspend fun startSession(): WearableStatus
    suspend fun disconnect()
    fun currentStatus(): WearableStatus
}

/**
 * Implementación de MODE B, fiel al flujo documentado oficialmente para
 * MockDeviceKit (ver docs/DOOMY_VISION_ARCHITECTURE.md#meta-wearables-dat):
 * "The device must be powered on and worn (donned) before streaming can
 * start." — replicamos esa regla aquí, no solo en la UI.
 */
class MockWearablesManager : WearablesManager {
    override val mode = DeviceMode.MOCK_DEVICE
    private var status = WearableStatus(WearableConnectionState.DISCONNECTED, cameraAvailable = false)

    override suspend fun pair(): WearableStatus {
        status = status.copy(connectionState = WearableConnectionState.PAIRED)
        return status
    }

    override suspend fun powerOn(): WearableStatus {
        require(status.connectionState >= WearableConnectionState.PAIRED) {
            "No se puede encender un dispositivo mock sin emparejar primero"
        }
        status = status.copy(connectionState = WearableConnectionState.POWERED_ON)
        return status
    }

    override suspend fun don(): WearableStatus {
        require(status.connectionState >= WearableConnectionState.POWERED_ON) {
            "El dispositivo debe estar encendido antes de 'ponérselo'"
        }
        status = status.copy(connectionState = WearableConnectionState.WORN)
        return status
    }

    override suspend fun unfold(): WearableStatus {
        require(status.connectionState >= WearableConnectionState.WORN) {
            "El dispositivo debe estar puesto (donned) antes de desdoblar"
        }
        return status
    }

    override suspend fun startSession(): WearableStatus {
        // Regla documentada de Meta: "must be powered on and worn (donned)
        // before streaming can start" — la aplicamos explícitamente.
        if (status.connectionState < WearableConnectionState.WORN) {
            throw DoomyVisionBridgeError.WearablesError(
                detail = "startSession() llamado sin power_on+don previos (estado actual: ${status.connectionState})"
            )
        }
        status = status.copy(connectionState = WearableConnectionState.SESSION_READY, cameraAvailable = true, batteryPercent = 87)
        return status
    }

    override suspend fun disconnect() {
        status = WearableStatus(WearableConnectionState.DISCONNECTED, cameraAvailable = false)
    }

    override fun currentStatus(): WearableStatus = status
}

// WearableConnectionState es un enum, por lo tanto ya implementa
// Comparable<WearableConnectionState> por orden de declaración (ordinal) —
// eso es lo que usan los `require(status.connectionState >= ...)` de arriba.

File: bridge-ios/Sources/DoomyVisionCore/Wearables.swift
import Foundation

/// Espejo de Wearables.kt — ver ese archivo para la justificación completa
/// (sección 6). MockWearablesManager reproduce el flujo documentado de
/// MockDeviceKit: pair -> power on -> don -> unfold -> session ready.
public enum DeviceMode { case realGlasses, mockDevice, phone }

public enum WearableConnectionState: Int, Comparable {
    case disconnected = 0, pairing, paired, poweredOn, worn, sessionReady
    public static func < (lhs: Self, rhs: Self) -> Bool { lhs.rawValue < rhs.rawValue }
}

public struct WearableStatus {
    public var connectionState: WearableConnectionState
    public var cameraAvailable: Bool
    public var batteryPercent: Int?
    public init(connectionState: WearableConnectionState, cameraAvailable: Bool, batteryPercent: Int? = nil) {
        self.connectionState = connectionState
        self.cameraAvailable = cameraAvailable
        self.batteryPercent = batteryPercent
    }
}

public protocol WearablesManager: AnyObject {
    var mode: DeviceMode { get }
    func pair() async throws -> WearableStatus
    func powerOn() async throws -> WearableStatus
    func don() async throws -> WearableStatus
    func unfold() async throws -> WearableStatus
    func startSession() async throws -> WearableStatus
    func disconnect() async
    func currentStatus() -> WearableStatus
}

public final class MockWearablesManager: WearablesManager {
    public let mode: DeviceMode = .mockDevice
    private var status = WearableStatus(connectionState: .disconnected, cameraAvailable: false)
    public init() {}

    public func pair() async throws -> WearableStatus {
        status.connectionState = .paired
        return status
    }

    public func powerOn() async throws -> WearableStatus {
        guard status.connectionState >= .paired else {
            throw DoomyVisionBridgeError.wearablesError(detail: "powerOn sin pair() previo")
        }
        status.connectionState = .poweredOn
        return status
    }

    public func don() async throws -> WearableStatus {
        guard status.connectionState >= .poweredOn else {
            throw DoomyVisionBridgeError.wearablesError(detail: "don() sin powerOn() previo")
        }
        status.connectionState = .worn
        return status
    }

    public func unfold() async throws -> WearableStatus {
        guard status.connectionState >= .worn else {
            throw DoomyVisionBridgeError.wearablesError(detail: "unfold() sin don() previo")
        }
        return status
    }

    public func startSession() async throws -> WearableStatus {
        // Regla documentada de Meta: powered on + worn antes de streaming.
        guard status.connectionState >= .worn else {
            throw DoomyVisionBridgeError.wearablesError(detail: "startSession sin power_on+don (estado: \(status.connectionState))")
        }
        status.connectionState = .sessionReady
        status.cameraAvailable = true
        status.batteryPercent = 87
        return status
    }

    public func disconnect() async {
        status = WearableStatus(connectionState: .disconnected, cameraAvailable: false)
    }

    public func currentStatus() -> WearableStatus { status }
}
